Gameplay/swimming exhaust

inxomnyaa opened the pull request:
## Introduction
<!-- Explain existing problems or why this pull request is necessary -->
Implements correct swimming exhaust as mentioned in #4668

### Relevant issues
<!-- List relevant issues here -->
<!--

* Fixes #1
* Fixes #2

-->
#4668

## Changes
### API changes
<!-- Any additions to the API that should be documented in release notes? -->
Added Entity::isInsideOfLiquid()

### Behavioural changes
<!-- Any change in how the server behaves, or its performance? -->
Swimming now uses the correct exhaust

## Backwards compatibility
<!-- Any possible backwards incompatible changes? How are they solved, or how can they be solved? -->

## Follow-up
<!-- Suggest any actions to be done before/after merging this pull request -->
<!--

Requires translations:

| Name | Value in eng.ini |
| :--: | :---: |
| `foo.bar` | `Foo bar` |

-->
It is to mention that `isUnderwater` has no way to consider if it should check for the boundingbox being fully submerged or partially in water, which might cause problems with things like fishing bobber (which in vanilla resets the RNG when it "bobs" out of water - it only checks for the head being submerged, as it is needed for breathing code.

~~Adding a isSubmergedInLiquid(Liquid, bool fully) might be a good idea for the future.~~ EDIT: isInsideOfLiquid() has been added to address this

## Tests
<!--
Details should be provided of tests done. Simply saying "tested" or equivalent is not acceptable.

Attach scripts or actions to test this pull request, as well as the result
-->
Since exhaust values are a bit wonky in PMMP anyways, the tests might be a bit flawed. It appears swimming is working fine though with the exhaust now, using up far less hunger than it used to.

~~One thing i am not entirely sure of yet is if the isSwimming and "!isUnderwater" works properly together, still testing~~
EDIT: after excessive tests i can conclude that exhaust is applied correctly now, both with "bobbing" swimming and sprint-swimming

inxomnyaa committed:
Implement swimming exhaust #4668

dktapps commented on src/player/Player.php:
```diff
@@ -1213,9 +1213,14 @@ protected function processMostRecentMovements() : void{
-				//TODO: check swimming (adds 0.015 exhaustion in MCPE)
 				if($this->isSprinting()){
 					$this->hungerManager->exhaust(0.1 * $horizontalDistanceTravelled, PlayerExhaustEvent::CAUSE_SPRINTING);
+				}elseif($this->isSwimming()){
```
This won't be set if the player is just bobbing up and down on the surface of water.

dktapps commented:
`isUnderwater()` is not the same as `isInWater()`, which is what we would need for the surface swimming.

inxomnyaa committed:
Added isInsideOfLiquid method to address exhaust when not sprinting
As mentioned in https://github.com/pmmp/PocketMine-MP/pull/4677#discussion_r775658973 https://github.com/pmmp/PocketMine-MP/pull/4677#issuecomment-1001784995

inxomnyaa committed:
Satisfy phpstan

inxomnyaa committed:
Fix isInsideOfLiquid always failing, skip exhaust if not survival

ShockedPlot7560 commented on src/entity/Entity.php:
```diff
@@ -1128,11 +1130,41 @@ public function isUnderwater() : bool{
 
 	public function isInsideOfSolid() : bool{
-		$block = $this->getWorld()->getBlockAt((int) floor($this->location->x), (int) floor($y = ($this->location->y + $this->getEyeHeight())), (int) floor($this->location->z));
+		$block = $this->getWorld()->getBlockAt((int) floor($this->location->x), (int) floor($this->location->y + $this->getEyeHeight()), (int) floor($this->location->z));
```
unrelated change

inxomnyaa replied:
Should be a follow-up, agreed

ShockedPlot7560 commented on src/entity/Entity.php:
```diff
@@ -1128,11 +1130,41 @@ public function isUnderwater() : bool{
+		$collisionBlocks = $this->getBlocksAroundWithEntityInsideActions();
+		$blocks = array_filter($collisionBlocks, static function($block) use ($liquid, $y, $blockY, $submerged) : bool{
+			if($submerged && $block instanceof $liquid){
+				$f = ($blockY + 1) - ($block->getFluidHeightPercent() - 0.1111111);
```
What's this value 0.1111 ?

inxomnyaa replied:
https://github.com/pmmp/PocketMine-MP/blob/77db7a8941d3d91c7e452cb3ca469f59fbd8f4df/src/entity/Entity.php#L1123

ShockedPlot7560 commented on src/entity/Entity.php:
```diff
@@ -1128,11 +1130,41 @@ public function isUnderwater() : bool{
+		$y = $this->location->y + $this->getEyeHeight();
+		$blockY = floor($y);
+		if($checkEyePos){
+			return $this->getWorld()->getBlockAt((int) floor($this->location->x), (int) $blockY, (int) floor($this->location->z)) instanceof $liquid;
```
Why not using floorX, floorZ methods from vector ?

ShockedPlot7560 commented on src/entity/Entity.php:
```diff
@@ -1128,11 +1130,41 @@ public function isUnderwater() : bool{
+			return $block instanceof $liquid;
+		});
+		$countFiltered = count($blocks);
+		if($countFiltered === 0){
+			return false;
+		}elseif(!$submerged){
+			return true;
+		}elseif($countFiltered === count($collisionBlocks)){
+			return true;
+		}
+		return false;
```
this conditions can be reduced maybe with only one.

inxomnyaa commented:
I don't really touch PHP anymore these days (can't be bothered to loose my mind for minor edits), so i'll leave it here for anyone to claim.
